src/app-layer-pop3.c

```diff
@@ -0,0 +1,1080 @@
+#define POP3_PARSER_STATE_PARSING_MULTILINE_REPLY 0x08
+
+/* Various POP3 commands
+ * We currently have var-ified just STARTTLS and DATA, since we need to them
+ * for state transitions.  The rest are just indicate as OTHER_CMD.  Other
+ * commands would be introduced as and when needed */
+#define POP3_COMMAND_USER      1
+#define POP3_COMMAND_RETR      2
+/* not an actual command per se, but the mode where we accept the mail after
+ * DATA has it's own reply code for completion, from the server.  We give this
+ * stage a pseudo command of it's own, so that we can add this to the command
+ * buffer to match with the reply */
+#define POP3_COMMAND_DATA_MODE 4
+/* All other commands are represented by this var */
+#define POP3_COMMAND_OTHER_CMD 5
+
+SCEnumCharMap pop3_decoder_event_table[ ] = {
+    { "INVALID_REPLY",           POP3_DECODER_EVENT_INVALID_REPLY },
+    { "UNABLE_TO_MATCH_REPLY_WITH_REQUEST",
+      POP3_DECODER_EVENT_UNABLE_TO_MATCH_REPLY_WITH_REQUEST },
+    { "MAX_COMMAND_LINE_LEN_EXCEEDED",
+      POP3_DECODER_EVENT_MAX_COMMAND_LINE_LEN_EXCEEDED },
+    { "MAX_REPLY_LINE_LEN_EXCEEDED",
+      POP3_DECODER_EVENT_MAX_REPLY_LINE_LEN_EXCEEDED },
+    { "INVALID_PIPELINED_SEQUENCE",
+      POP3_DECODER_EVENT_INVALID_PIPELINED_SEQUENCE },
+    { "BDAT_CHUNK_LEN_EXCEEDED",
+      POP3_DECODER_EVENT_BDAT_CHUNK_LEN_EXCEEDED },
+    { "NO_SERVER_WELCOME_MESSAGE",
+      POP3_DECODER_EVENT_NO_SERVER_WELCOME_MESSAGE },
+    { "TLS_REJECTED",
+      POP3_DECODER_EVENT_TLS_REJECTED },
+    { "DATA_COMMAND_REJECTED",
+      POP3_DECODER_EVENT_DATA_COMMAND_REJECTED },
+
+    /* MIME Events */
+    { "MIME_PARSE_FAILED",
+      POP3_DECODER_EVENT_MIME_PARSE_FAILED },
+    { "MIME_MALFORMED_MSG",
+      POP3_DECODER_EVENT_MIME_MALFORMED_MSG },
+    { "MIME_INVALID_BASE64",
+      POP3_DECODER_EVENT_MIME_INVALID_BASE64 },
+    { "MIME_INVALID_QP",
+      POP3_DECODER_EVENT_MIME_INVALID_QP },
+    { "MIME_LONG_LINE",
+      POP3_DECODER_EVENT_MIME_LONG_LINE },
+    { "MIME_LONG_ENC_LINE",
+      POP3_DECODER_EVENT_MIME_LONG_ENC_LINE },
+    { "MIME_LONG_HEADER_NAME",
+      POP3_DECODER_EVENT_MIME_LONG_HEADER_NAME },
+    { "MIME_LONG_HEADER_VALUE",
+      POP3_DECODER_EVENT_MIME_LONG_HEADER_VALUE },
+
+    { NULL,                      -1 },
+};
+
+typedef struct POP3Config {
+
+    int decode_mime;
+    MimeDecConfig mime_config;
+    uint32_t content_limit;
+    uint32_t content_inspect_min_size;
+    uint32_t content_inspect_window;
+
+    StreamingBufferConfig sbcfg;
+} POP3Config;
+
+/* Create POP3 config structure */
+static POP3Config pop3_config = { 0, { 0, 0, 0, 0, 0 }, 0, 0, 0, STREAMING_BUFFER_CONFIG_INITIALIZER };
+
+/**
+ * \brief Configure POP3 Mime Decoder by parsing out 'pop3-mime' section of YAML
+ * config file
+ *
+ * \return none
+ */
+static void POP3Configure(void) {
+
+    SCEnter();
+    int ret = 0, val;
+    intmax_t imval;
+
+    ConfNode *config = ConfGetNode("app-layer.protocols.pop3.mime");
+    if (config != NULL) {
```

Comment:
Seems this is a duplicate of the SMTP code, which is also duplicated to IMAP. Can we do this as a util func instead?